### PR TITLE
Ensure no overlap between SDN cluster network and service/portal network

### PR DESCRIPTION
**Pull request by pravisankar**

### CODE REVIEW

**Comment by pravisankar** [comment]
@danwinship PTAL

File: pkg/sdn/api/validation/validation.go [hunk 1]
```diff
@@ -15,21 +15,25 @@ func ValidateClusterNetwork(clusterNet *sdnapi.ClusterNetwork) fielderrors.Valid
...
 		}
 	}
 
-	_, _, err = net.ParseCIDR(clusterNet.ServiceNetwork)
+	serviceIP, _, err := net.ParseCIDR(clusterNet.ServiceNetwork)
 	if err != nil {
 		allErrs = append(allErrs, fielderrors.NewFieldInvalid("serviceNetwork", clusterNet.ServiceNetwork, err.Error()))
 	}
 
+	if clusterIPNet.Contains(serviceIP) {
```
Comment (danwinship) [inline pkg/sdn/api/validation/validation.go]: You need to test if serviceIPNet.Contains(clusterIP) as well

Comment (liggitt) [inline pkg/sdn/api/validation/validation.go]: what will this do to existing data?

Comment (pravisankar) [inline pkg/sdn/api/validation/validation.go]: You mean existing cluster with overlapping cluster/service network?
If that is the case, then the cluster is already messed up..openshift-node won't be able to differentiate between pod and service traffic.

**Comment by pravisankar** [comment]
@danwinship updated

File: pkg/sdn/api/validation/validation.go (hunk 1 above)
Comment (liggitt) [inline pkg/sdn/api/validation/validation.go]: And the things that would get an update error would handle that well, and the configuration error could be corrected?

Comment (liggitt) [inline pkg/sdn/api/validation/validation.go]: Can you check with the install team to make sure the same validation is done when setting up the configuration?

Comment (pravisankar) [inline pkg/sdn/api/validation/validation.go]: On Thu, Oct 22, 2015 at 3:14 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/sdn/api/validation/validation.go
> <https://github.com/openshift/origin/pull/5345#discussion_r42813410>:
>
> >  	if err != nil {
> >  		allErrs = append(allErrs, fielderrors.NewFieldInvalid("serviceNetwork", clusterNet.ServiceNetwork, err.Error()))
> >  	}
> >
> > +	if clusterIPNet.Contains(serviceIP) {
>
> And the things that would get an update error would handle that well, and
> the configuration error could be corrected?
>
Yes. Another related issue
https://github.com/openshift/openshift-sdn/issues/194

> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openshift/origin/pull/5345/files#r42813410>.
>



-- 
-Ravi

Comment (pravisankar) [inline pkg/sdn/api/validation/validation.go]: Duplicate check in ansible scripts to fail fast?
cc: @sdodson

**Comment by danwinship** [comment]
lgtm

File: pkg/sdn/api/validation/validation.go (hunk 1 above)
Comment (sdodson) [inline pkg/sdn/api/validation/validation.go]: @detiber Do you know of a way to do subnet calculations in ansible code or would we have to bury this in openshift_facts?

**Comment by pravisankar** [comment]
Updated: Added test cases

**Comment by danwinship** [comment]
[merge]

**Comment by dcbw** [comment]
LGTM

**Comment by smarterclayton** [comment]
[merge]

**Pull request merged**